### ISSUE TITLE
Bug: fail to throw error when computing loss between tensors with shapes [n, 1] and [n]

### ISSUE THREAD

**Issue by rmurphy2718** [pytorch/pytorch#16045]
## Issue description
When computing the loss between ```predicted``` and ```target``` with shapes such as ```target.shape = [100]``` and ```predicted.shape = [100, 1]```, an error should be thrown.  Indeed, my colleagues tell me this was the case in version 0.4, but lost in version 1.0.

## Code example
```
# Version 1.0
# Set up straightforward linear regression problem.
import torch
import torch.nn as nn
sample_size = 100
x = torch.randn((sample_size,2))
eps = torch.randn(sample_size )
y = 10 + 9*x[:,0] - 2*x[:, 1] +  eps
#y = y.reshape(sample_size, 1)
model = nn.Linear(2, 1)

optimizer = torch.optim.SGD(model.parameters(), lr=.1)
loss_func = nn.MSELoss()
for epoch in range(30):
   optimizer.zero_grad()
   yhat = model(x)
   loss_val = loss_func(yhat, y)
   loss_val.backward()
   optimizer.step()
   print(loss_val.item())

print(model.weight.data)
print(model.bias.data)

```
If you leave y.reshape commented out, the correct estimates of this straightforward linear regression problem are not found and the loss fails to decrease into a region near the global optimum of this convex problem.  Yet, no error is raised.

If you uncomment y.reshape, the correct estimates are found and the loss decreases in the anticipated way.

## System Info
Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.5
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl  
[conda] mkl                       2017.0.3                      0  
[conda] torch                     1.0.0                     <pip>
[conda] torchvision               0.2.1                     <pip>

**Comment by mkakodkar** [comment]
Same behavior on 
Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14
GCC version: Could not collect
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

**Comment by zou3519** [comment]
MSELoss does broadcasting between tensors of shape `(n, 1)` and `(n)`, resulting in an output of shape `(n, n)` which is probably not what you wanted.

**Comment by rmurphy2718** [comment]
In my humble opinion, this behavior is sufficiently unexpected to warrant throwing an error or warning.

**Comment by cosmin-pascaru** [comment]
I have encountered this exact problem, and debugging it was very time consuming, I believe a warning at least would be very important. 

If this mistake is made in the first iterations on a model and is not noticed, it might invalidate some ideas all together.

**Comment by zou3519** [comment]
Let's say a user writes their MSELoss without using the nn.Module, as 
```
def loss(x, t):
   torch.sum((x - t) ** 2) / x.numel()
```
Would you expect this function to warn against inputs of size `n` and `n, 1`?

**Comment by cosmin-pascaru** [comment]
> Let's say a user writes their MSELoss without using the nn.Module, as
> 
> ```
> def loss(x, t):
>    torch.sum((x - t) ** 2) / x.numel()
> ```
> Would you expect this function to warn against inputs of size `n` and `n, 1`?

Indeed, this is a case that cannot be caught, without a way of disabling broadcasting globally through a config setting, or through the use of a context manager, though I assume that would be hard to implement and quite error-prone. 

But I would argue that the sole purpose of `MSELoss` (and other loss functions) is to compute a loss between 2 theoretical same-shape tensors, and so the cases where broadcasting would be useful is limited. I cannot think of a real-world, useful example for such scenario. 

I would say that some warnings are better than no warnings.
For example, tensorflow is performing checks in such cases, for example: https://github.com/tensorflow/tensorflow/issues/1406

**Comment by zou3519** [comment]
cc @ailzhang @fmassa do you remember why we decided to let these losses broadcast?

**Comment by ptrblck** [comment]
+1 to throw a warning/error for a shape mismatch.
We had a few threads in the past about bad training when the tensors were silently broadcasted using `nn.MSELoss`.
[Example1](https://discuss.pytorch.org/t/sklearn-model-to-pytorch-model/33532/3?u=ptrblck)
[Example2](https://discuss.pytorch.org/t/pytorch-beginner-model-fit-problems/37292/2?u=ptrblck)

**Comment by mc-robinson** [comment]
Having spent way too long finding this error, I would be happy to submit a pull request -- but I would appreciate your help/thoughts on a few considerations first.

1. I can't think of any case in which broadcasting is useful for MSELoss(). Therefore, throwing an error is seemingly more appropriate to me. However, I've often seen training functions like this:

```python
def train_epoch(model:nn.Module, dl:DataLoader, opt:optim.Optimizer, loss_func:LossFunction)->None:
    "Simple training of `model` for 1 epoch of `dl` using optim `opt` and loss function `loss_func`."
    model.train()
    for xb,yb in dl:
        loss = loss_func(model(xb), yb)
        loss.backward()
        opt.step()
        opt.zero_grad()
```

from fastai https://github.com/fastai/fastai/blob/5ac8dfe800dddb6ede6ce4178a1d68de7b087f6e/fastai/basic_train.py . `model(xb)` has shape (N,1) while `yb` has shape (N,). In this case, the DataLoader is returning the shapes that cause this function not to work (I am also submitting an issue to fastai). Therefore, returning an error here every time seems a bit harsh unless DataLoader is changed.

2. Considering (1), I might propose something like the following fix in F.mse_loss() (adapted from the warnings/errors in F.binary_cross_entropy()). I am going off the assumption that anytime one inputs a tensor of dimension 1, they do not want to do any special broadcasting. This seems fair to me, but feel free to disagree -- I have never contributed here before and definitely would appreciate comments.

```python
if not (target.size() == input.size()):
    # first handle the one-dimensional cases 
    if target.ndimension() == 1:
        target = target.unsqueeze(1)
    elif input.ndimension() == 1:
        input = input.unsqueeze(1)
    else:
        warnings.warn("Using a target size ({}) that is different to the input size ({})."
                      "This will likely lead to incorrect results due to broadcasting. "
                      "Please ensure they have the same size.".format(target.size(), input.size()))
        
if input.numel() != target.numel():
    raise ValueError("Target and input must have the same number of elements. target nelement ({}) "
                     "!= input nelement ({})".format(target.numel(), input.numel()))
```

**Comment by ailzhang** [comment]
Hmmm I don't remember why we decided to broadcast for some loss functions, it might be when some of the loss functions were implemented in python, they got broadcast for free and people start seeing mismatched behaviors. 
I'm fine with throwing errors on this. 
cc: @gchanan  @fmassa your thoughts?

**Comment by soumith** [comment]
we should throw a warning when broadcasted for nn.MSELoss i think.

**Comment by soumith** [comment]
i'm also supportive of a hard error, but a warning would be sufficient

**Comment by ailzhang** [comment]
@mc-robinson  Feel free to submit a PR if you are interested!

**Comment by mc-robinson** [comment]
Ok will do! Do we think my code was alright or do we just want to stick with a warning for now before broadcasting?

**Comment by mc-robinson** [comment]
Hard error has the potential to impact fair amount of training code... For example, the following code would lose its generality unless we deal with DataLoader shapes automatically in the loss function

```python
model.train()
for xb,yb in dl:
        loss = loss_func(model(xb), yb)
        loss.backward()
        opt.step()
        opt.zero_grad()
```

**Comment by ailzhang** [comment]
Yea I think we might prefer a big warning at this moment, and make it a hard error in a future release when users are used to this behavior. A hard BC breaking is bad I guess.

**Comment by mc-robinson** [comment]
Ok, I'll at least get a warning out there for now

**Comment by gchanan** [comment]
warnings without a way to turn them off are pretty annoying.  You could do something like add two parameters:
`warn_broadcasting=True, error_broadcasting=False`.

which would allow us to support all the combinations pretty easily but still have backwards compatibility and still give the user control.

**Comment by gchanan** [comment]
closing for now since the warning has been implemented.  If there is sufficient request for an error option, please open another issue.

**Comment by lkhphuc** [comment]
Just wondering is there any reason this warning is not implemented for other (or all) losses?
I just spend an awful lot of time on this for the L1Loss.
